rclcpp/test/benchmark/benchmark_init_shutdown.cpp

```diff
@@ -0,0 +1,41 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "performance_test_fixture/performance_test_fixture.hpp"
+
+#include "rclcpp/rclcpp.hpp"
+
+using performance_test_fixture::PerformanceTest;
+
+BENCHMARK_F(PerformanceTest, rclcpp_init)(benchmark::State & state)
+{
+  for (auto _ : state) {
+    rclcpp::init(0, nullptr);
+
+    state.PauseTiming();
+    rclcpp::shutdown();
+    state.ResumeTiming();
+  }
+}
+
+BENCHMARK_F(PerformanceTest, rclcpp_shutdown)(benchmark::State & state)
+{
+  for (auto _ : state) {
+    state.PauseTiming();
+    rclcpp::init(0, nullptr);
+    state.ResumeTiming();
+
+    rclcpp::shutdown();
+  }
```

Comment:
The `PauseTiming` and `ResumeTiming` concepts are evidently pretty expensive and can make the measurements noisy. The approach we've been taking in the other benchmarks is to time the cycle (in this case, startup/shutdown, elsewhere init/fini, etc).

That said, I haven't actually tried to measure how bad Pause/Resume makes things. If you can gather some data on the subject, I could be convinced to start using them, otherwise I think we should opt for the less noisy approach, since the EC2 instances are generating enough noise as it is.

Reply:
@cottsay one downside to using these Pause/Resume calls is that the benchmark library will require running many more iterations to get much higher certainty in the measurements. But if the code inside the Pause/Resume is slow, then the benchmark can take a long time to run. The prime example here is `create_node` in `test_node__rmw_connext_cpp`. There creating the node is fast, but destroying it is very slow (3.18 seconds) so benchmark wants to run that test a many many times, instead of just once, which it does for the `destroy_node` benchmark (and the new `create_destroy_node` one).  Ultimately, this caused the benchmark to timeout.

I think the current default is that it tries to collect about 1 second worth of iterations? If Pause/Resume is then used and the code outside of Pause/Resume is 50% of the total iteration runtime, the benchmark would take 2 seconds to run. If that code is an even smaller percentage it would take many more seconds to run. This will easily lead to longer benchmark runtimes if not used carefully for rclcpp.

Therefore my recommendation is that if the code run outside of the Pause/Resume calls is small (< 5 us) Pause/Resume should be avoided. OR if the benchmark is not critical and if the code outside Pause/Resume is less than 50% of the runtime of the iteration, then it should also be avoided.